text/0000-inferred-type-aliases.md

```diff
@@ -0,0 +1,696 @@
+let bar: usize = foo;
+```
+
+### `opaque`
+[opaque]: #opaque
+
+The opposite of something being transparent is for something to be *"opaque"*.
+This means that you can *not* see through it. In the context of Rust,
+this refers to that the true nominal type of some object or expression
+is hidden from the user. Instead only a particular interface is exposed.
+This entails that if you have such an expression, only the operations
+afforded by the interface may be performed. This has been provided by
+the `impl Trait` (see [RFC 1522], [RFC 1951], [RFC 2071], [RFC 2250],
+and currently proposed [RFC 2515]) mechanism in Rust.
+
+An example:
+
+```rust
+use std::ops::Range;
+
+fn one_to_hundred() -> impl Iterator<Item = u8> {
+    let range: Range<u8> = 0..100;
+    range
+}
+
+fn main() {
+    let _numbers: Range<u8> = one_to_hundred();
+}
+```
+
+This results in:
+
+```rust
+error[E0308]: mismatched types
+ --> src/main.rs:9:30
+  |
+9 |     let _numbers: Range<u8> = one_to_hundred();
+  |                               ^^^^^^^^^^^^^^^^ expected struct `std::ops::Range` found anonymized type
+  |
+  = note: expected type `std::ops::Range<u8>`
+             found type `impl std::iter::Iterator`
+
+error: aborting due to previous error
+```
+
+Here *"anonymized"* and *"opaque"* mean the same thing.
+As we can see from the error message, even though we know that the nominal type
+of `_numbers` is `Range<u8>`, the compiler has prevented us from concluding this.
+
+### `voldemort type`
+[voldemort type]: #voldemort-type
+
+A *"voldemort type"* in Rust refers to types which *can not be named*.
+They are also called *"unnameable types"*. The canonical example of such
+unnameable types are the types of closures. Each closure has a distinct type,
+which the compiler knows about, but you can't write their name in code.
+
+An example:
+
+```rust
+#![feature(core_intrinsics)]
+
+use std::intrinsics::type_name;
+
+fn name_of_type<T>(_: &T) -> &str {
+    unsafe { type_name::<T>() }
+}
+
+fn main() {
+    let closure_one = |x: usize| x + 1;
+    let closure_two = |x: usize| x + 1;
+    println!("{}", name_of_type(&closure_one));
+    // output: [closure@src/main.rs:10:23: 10:39]
+    println!("{}", name_of_type(&closure_two));
+    // output: [closure@src/main.rs:11:23: 11:39]
+}
+```
+
+As you can see from the output, the type of these two closures are *not* the same.
+
+## The Proposal
+[the proposal]: #the-proposal
+
+Now that we've defined the core concepts this RFC deals with,
+let's deal with what is actually proposed.
+
+This part is rather simple.
+What this RFC proposes is that you should be able to have the actual type of a
+`type` alias as well as the associated type of an implementation be [inferred]
+and that these inferred types should be [transparent] as opposed to [opaque].
+We propose to do this by using the familiar symbol `_` which means
+*"please infer the type"* in type contexts.
+
+For example, you will be allowed to write:
```

Comment:
These examples all have the type being inferred from the function bodies, like RPIT, but presumably `_` would allow inference from solely the caller too. Can you make sure there are examples of that, if you haven't already?